[Style] Use standard Blending interface for shadow interpolation

weinig opened the pull request:
#### 81c9581356634b90e2ef020950bcce0f00b13fc4
<pre>
[Style] Use standard Blending interface for shadow interpolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=299193">https://bugs.webkit.org/show_bug.cgi?id=299193</a>

Reviewed by Darin Adler.

Moves shadow interpolation code to new file StyleShadowInterpolation.h
and uses that from `box-shadow` and `text-shadow` Blending interfaces,
moving the default shadow accessor used by the shared blending code
to the appropriate files as well.

Adds convenience constructors to FixedVector/RefCountedFixedVector
wrappers for indexed construction to allow removing a bit of duplication.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
* Source/WebCore/style/values/borders/StyleBoxShadow.h:
* Source/WebCore/style/values/borders/StyleShadowInterpolation.h: Added.
* Source/WebCore/style/values/svg/StyleSVGStrokeDasharray.cpp:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.h:
* Source/WebCore/style/values/transforms/StyleTransformList.cpp:
* Source/WebCore/style/values/ui/StyleCursor.cpp:

Canonical link: <a href="https://commits.webkit.org/300280@main">https://commits.webkit.org/300280@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db03f8640ff50d82582769c4d0844e2ad218bfc1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122012 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41714 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32384 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128572 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74105 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123888 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50308 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92731 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33832 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73385 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32845 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27425 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72069 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103339 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27621 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131336 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48951 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/37224 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/101290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49325 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105478 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101161 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25648 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46531 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24648 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45660 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48808 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54542 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48278 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49958 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp:
```diff
@@ -106,5 +108,29 @@ auto Blending<TextShadow>::blend(const TextShadow& a, const TextShadow& b, const
+struct MatchingTextShadows {
+    static const TextShadow& shadowForInterpolation(const TextShadow&)
+    {
+        static NeverDestroyed<TextShadow> defaultShadowData {
```
```suggestion
        static NeverDestroyed<const TextShadow> defaultShadowData {
```

darinadler commented on Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
```diff
@@ -127,5 +129,66 @@ auto Blending<BoxShadow>::blend(const BoxShadow& a, const BoxShadow& b, const Re
+struct MatchingBoxShadows {
+    static const BoxShadow& shadowForInterpolation(const BoxShadow& shadowToMatch)
+    {
+        static NeverDestroyed<BoxShadow> defaultShadowData {
```
```suggestion
        static NeverDestroyed<const BoxShadow> defaultShadowData {
```

darinadler commented on Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
```diff
@@ -127,5 +129,66 @@ auto Blending<BoxShadow>::blend(const BoxShadow& a, const BoxShadow& b, const Re
+                .isWebkitBoxShadow = false
+            }
+        };
+        static NeverDestroyed<BoxShadow> defaultInsetShadowData {
```
```suggestion
        static NeverDestroyed<const BoxShadow> defaultInsetShadowData {
```

darinadler commented on Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
```diff
@@ -127,5 +129,66 @@ auto Blending<BoxShadow>::blend(const BoxShadow& a, const BoxShadow& b, const Re
+                .isWebkitBoxShadow = false
+            }
+        };
+        static NeverDestroyed<BoxShadow> defaultWebKitBoxShadowData {
```
```suggestion
        static NeverDestroyed<const BoxShadow> defaultWebKitBoxShadowData {
```

darinadler commented on Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
```diff
@@ -127,5 +129,66 @@ auto Blending<BoxShadow>::blend(const BoxShadow& a, const BoxShadow& b, const Re
+                .isWebkitBoxShadow = true
+            }
+        };
+        static NeverDestroyed<BoxShadow> defaultInsetWebKitBoxShadowData {
```
```suggestion
        static NeverDestroyed<const BoxShadow> defaultInsetWebKitBoxShadowData {
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/db03f8640ff50d82582769c4d0844e2ad218bfc1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51018

weinig committed:
[Style] Use standard Blending interface for shadow interpolation
https://bugs.webkit.org/show_bug.cgi?id=299193

Reviewed by Darin Adler.

Moves shadow interpolation code to new file StyleShadowInterpolation.h
and uses that from `box-shadow` and `text-shadow` Blending interfaces,
moving the default shadow accessor used by the shared blending code
to the appropriate files as well.

Adds convenience constructors to FixedVector/RefCountedFixedVector
wrappers for indexed construction to allow removing a bit of duplication.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/style/StyleInterpolationFunctions.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
* Source/WebCore/style/values/borders/StyleBoxShadow.h:
* Source/WebCore/style/values/borders/StyleShadowInterpolation.h: Added.
* Source/WebCore/style/values/svg/StyleSVGStrokeDasharray.cpp:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.h:
* Source/WebCore/style/values/transforms/StyleTransformList.cpp:
* Source/WebCore/style/values/ui/StyleCursor.cpp:

Canonical link: https://commits.webkit.org/300280@main
